### PR TITLE
Transformation services extensions

### PR DESCRIPTION
**Pull request by clinique**
=> Ported Scale transformation service to ESH
   * enhanced scale definition capacity using Ranges
=> Factorization of Cache system from Map transformation with Scale transformation in superclass (AbstractFileTransformationService)
   * enabled cache system to operate properly on suddirectories
   * ported file localization support as exists in OH1
=> Ported JSonPath transformation service to ESH

=> Small warning supression in existing tests
=> Added JUnit tests for Scale, Map and JSonPath

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Hi @clinique, thanks a lot for porting this code! I assume that it has been already reviewed in openHAB 1, so I won't have to go into too many details?
As you have added new libraries, you will also need to update the about.html file (see https://www.eclipse.org/legal/epl/about.php). I will also have to request approval from the IP team for them, before I can merge.

**Comment by clinique** [comment]
Hi @kaikreuzer ,
Yes, it is currently merged in OH1 but I made a revamp of Scale transformation in order to enhance it a bit and share localization and the cache ability you introduced in Map transformation in a superclass.
Then... I wouldn't assume my code does not need a review :) made my best to have the most commented and clean as possible, but I'm not superman.
Will take action toward about.html (libraries are under Apache license).

**Comment by clinique** [comment]
Squashed my commits

**Comment by kaikreuzer** [comment]
> As you have added new libraries, you will also need to update the about.html file

This is still missing, right?

**Comment by clinique** [comment]
Done

**Comment by kaikreuzer** [comment]
Sorry @clinique for not yet further processing this. As mentioned in #340, I would actually suggest to refactor this into separate bundles, so that users do not have to include all transform dependencies, but only the ones that they want to use. Let's first wait until #340 is merged and then it would be great if you could refactor this contribution accordingly - sorry for the inconvenience, but it is all for keeping a clean architecture on the long run :-)

**Comment by clinique** [comment]
@kaikreuzer , no problem, there is no hurry and keeping the project clean is key

File: bundles/core/org.eclipse.smarthome.core.transform.test/src/test/java/org/eclipse/smarthome/core/transform/internal/JSonPathTransformationServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.eclipse.smarthome.core.transform.TransformationException;
+import org.eclipse.smarthome.core.transform.internal.service.JSonPathTransformationService;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author Gaël L'hopital
+ */
+public class JSonPathTransformationServiceTest extends AbstractTransformationServiceTest {
```
Comment (sja) [inline bundles/core/org.eclipse.smarthome.core.transform.test/src/test/java/org/eclipse/smarthome/core/transform/internal/JSonPathTransformationServiceTest.java]: Would be nice to see here a failure test which produces a TransformationException as well.
Maybe you can copy it from my PR: https://github.com/sja/smarthome/commit/003f7218739ec1912bf4151dfe15f8c937a8d12a

File: bundles/core/org.eclipse.smarthome.core.transform/src/main/java/org/eclipse/smarthome/core/transform/internal/service/JSonPathTransformationService.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ 			throw new TransformationException("the given parameters 'JSonPath' and 'source' must not be null");
+ 		}
+
+ 		logger.debug("about to transform '{}' by the function '{}'", source, jsonPathExpression);
+
+ 		try {
+ 			Object transformationResult = JsonPath.read(source, jsonPathExpression);
+ 			logger.debug("transformation resulted in '{}'", transformationResult);
+ 			return (transformationResult != null) ? transformationResult.toString() : null;
+ 		} catch(InvalidPathException e) {
```
Comment (sja) [inline bundles/core/org.eclipse.smarthome.core.transform/src/main/java/org/eclipse/smarthome/core/transform/internal/service/JSonPathTransformationService.java]: Won't you difference between PathNotFoundException and InvalidCriteriaException and InvalidPathException? I think if a path expression won't match, it's not an error. In some cases it's ok to have no match and a `return null` would be expressive enough to get that.
InvalidCriteria and InvalidPath should result in config error for the related thing.
See (my commit)[https://github.com/sja/smarthome/commit/003f7218739ec1912bf4151dfe15f8c937a8d12a] and tell me, if you agree to this or think your solution is more correct.

Comment (clinique) [inline bundles/core/org.eclipse.smarthome.core.transform/src/main/java/org/eclipse/smarthome/core/transform/internal/service/JSonPathTransformationService.java]: Yes, I agree with your proposal.
Feel free to fork it and make a PR with complementary exception handling, as well as your nice test unit. Won't have time to do it myself

**Comment by sja** [comment]
Ok, I'll fork it as you suggested. As far as I saw, the changes of #340 have to be adopted as well, which I can (have to) do as well. :-)

**Comment by clinique** [comment]
Yes, I guess it will lengthen a bit the merging process, but from the ground it's a good point.

**Comment by kaikreuzer** [comment]
@sja Yes, it could make sense to base your PR on https://github.com/eclipse/smarthome/pull/340 - I am only waiting for approval by the IP team before I can merge this.

**Comment by kaikreuzer** [comment]
Fyi: #340 has just been merged!

**Comment by clinique** [comment]
Ok, I will close this PR and make two distinct integrating #340 : one for Localizable file along with Scale transformation service and another with JSonPath transformation service

**Pull request closed without merging**